Update graphql, events & extending docs for Laravel and v4 (#2026)
api-platform/docs: core/graphql.md

## Enabling GraphQL

To enable GraphQL and its IDE (GraphiQL and GraphQL Playground) in your API, simply require the [API Platform GraphQL](https://github.com/api-platform/graphql) package using Composer:

```console
    composer require api-platform/graphql
```

You can now use GraphQL at the endpoint: `https://localhost:8443/graphql`.

> [!NOTE]
> If you used [the Symfony Variant thanks to Symfony Flex](../symfony/index.md#using-symfony-flex-and-composer-advanced-users) or the Laravel variant, URLs will be prefixed with `/api` by default. For example, the GraphQL endpoint will be: `https://localhost:8443/api/graphql`.

## Changing Location of the GraphQL Endpoint
Sometimes you may want to have the GraphQL endpoint at a different location. This can be done by manually configuring the GraphQL controller.

> ### Symfony config for routes

Reviewer: ```suggestion
### Symfony Config for Routes
```

> ### Symfony config for routes

Reviewer: ```suggestion
### Symfony Routes
```

> Using the Symfony variant we can do this modification by adding the following code:

Reviewer: ```suggestion
Using the Symfony variant we can do this modification by adding the following code:

```

## Enabling GraphQL

> To enable GraphQL and its IDE (GraphiQL and GraphQL Playground) in your API, simply require the [API Platform GraphQL](https://github.com/api-platform/graphql) package using Composer:

Reviewer 2: ```suggestion
To enable GraphQL and its IDE (GraphiQL and GraphQL Playground) in your API, simply require the `api-platform/graphql` package using Composer:
```

Change `/api/graphql` to the URI you wish the GraphQL endpoint to be accessible on.

> ### Laravel config for routes

Reviewer: ```suggestion
### Laravel Routes
```

Using the Laravel variant we can do this modification by adding the following code:
```php
// routes/web.php
use Illuminate\Support\Facades\Route;
use ApiPlatform\GraphQL\Action\EntrypointAction;

Route::post('/api/graphql', EntrypointAction::class)
    ->name('api_graphql_entrypoint');
```

Change `/api/graphql` to the URI you wish the GraphQL endpoint to be accessible on.

## GraphiQL

Go to the GraphQL endpoint with your browser, you will see a nice interface provided by GraphiQL to interact with your API.

The GraphiQL IDE can also be found at `/graphql/graphiql`.

If you need to disable it, it can be done in the configuration:

> ### Disable GraphiQL with Symfony

Reviewer: ```suggestion
### Disabling GraphiQL with Symfony
```

# ...
```

> ### Disable GraphiQL with Laravel

Reviewer: ```suggestion
### Disabling GraphiQL with Laravel

```

# ...
```

### Disabling GraphiQL with Laravel

```php
<?php

> //api/config/api-platform.php

Reviewer 2: ```suggestion
// app/config/api-platform.php
```

I think this is not inside `api`, can you check?

# ...
```

### Disable GraphQL Playground with Laravel

```php
<?php
//api/config/api-platform.php
return [
    // ....
    'graphql' => [
        'graphql_playground' => [

>             'enabled' => false,

Reviewer 2: this is not yet supported inside Laravel we should add it (basically we only need to register the controller)

}
```

### Custom Queries config for Symfony

If you use autoconfiguration (the default Symfony configuration) in your application, then you are done!

> Else, you need to tag your resolver like this if you using Symfony without autoconfiguration :

Reviewer 3: ```suggestion
Else, you need to tag your resolver like this if you are using Symfony without autoconfiguration :
```

### Subscriptions Cache

Internally, API Platform stores the subscriptions in a cache, using the [Symfony Cache](https://symfony.com/doc/current/cache.html) if you're using the Symfony variant

> or ÆPI Platform uses [Laravel cache](https://laravel.com/docs/cache) if you're use the Laravel variant.

Reviewer 3: ```suggestion
or API Platform uses [Laravel cache](https://laravel.com/docs/cache) if you're using the Laravel variant.
```

Even if you create your [custom queries](#custom-queries) or your [custom mutations](#custom-mutations),
these resolvers will be used and yours will be called at the right time.

See the [Extending API Platform](extending.md) documentation for more information.

### Disabling system providers and processors

> If you need to, you can disable some states providers and state processors done by the resolvers, for instance if you don't want your data to be validated.

Reviewer 2: ```suggestion
If you need to, you can disable some states providers and state processors, for instance if you don't want your data to be validated.
```

};
```

#### Register the Error handler using Laravel

```php
<?php

namespace App\Providers;

use App\Error\ErrorHandler as ErrorHandlerDecorated;
use ApiPlatform\GraphQl\Error\ErrorHandler;
use Illuminate\Contracts\Foundation\Application;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    public function register(): void
    {

>         $this->app->singleton(ErrorHandlerDecorated::class, function (Application $app) {
>             return new ErrorHandlerDecorated();
>         });
>         

Reviewer 2: ```suggestion
```

        $this->app->extend(ErrorHandler::class, function (ErrorHandler $errorHandler, Application $app) {
            return new ErrorHandlerDecorated($errorHandler);
        });
    }

> 
>     public function boot(): void
>     {
>     }

Reviewer 2: ```suggestion
```

#### Custom Exception Normalizer

> > [!CAUTION]
> > Works only for API Platform with Symfony
> 

Reviewer 2: ```suggestion
```

#### Custom Exception Normalizer

> [!CAUTION]

> > Works only for API Platform with Symfony,

Reviewer 3: ```suggestion
> Works only for API Platform with Symfony
```

To use it please [modify the extracted types](#modify-the-extracted-types) or use it directly in [custom queries](#custom-queries) or [custom mutations](#custom-mutations).

### Custom Types config for Laravel

If you are using Laravel tag your type with:

```php
<?php

namespace App\Providers;

use ApiPlatform\GraphQl\Type\Definition\TypeInterface;
use App\Type\Definition\DateTimeType
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    public function register(): void
    {
        $this->app->tag([DateTimeType::class], TypeInterface::class);
    }

> 
>     public function boot(): void
>     {
>     }

Reviewer 2: ```suggestion
```

Really not sure but I think it's more readable no?

}
```

Your custom type is now registered and is available in the `TypesContainer`.

To use it please [modify the extracted types](#modify-the-extracted-types) or use it directly in [custom queries](#custom-queries) or [custom mutations](#custom-mutations).

## Modify the Extracted Types

> [!WARNING]
> This part is only supported for API Platform with Symfony

> 

Reviewer 2: we can extend that using laravel as well ?

All `DateTimeInterface` properties will have the `DateTime` type in this example.

## Changing the Serialization Context Dynamically (only for Symfony)

> [!WARNING]

> > This part is only supported for API Platform with Symfony

Reviewer 2: this also works for laravel

## Export the Schema in SDL

> [!WARNING]

> > These commands are only available for API Platform with Symfony!

Reviewer 2: ```suggestion
> This command is not yet available with Laravel, you're welcome to contribute [on Github](github.com/api-platform/core)
```
